project2/kernel.c

```diff
@@ -268,6 +282,114 @@ void Kernel_Request_Terminate() {
+    // Periodic tasks cannot send
+    if (Cp->priority == PERIODIC) {
+        OS_Abort(PERIODIC_MSG);
+        return;
+    }
+
+    // Check if process id is valid
+    if (!VALID_ID(request_info->msg_to)) {
+        OS_Abort(INVALID_REQ_INFO);
+        return;
+    }
+
+    PD *p_recv = &Process[request_info->msg_to];
+    MTYPE recv_mask = p_recv->req_params->msg_mask;
+
+    // Check if info.msg_to is waiting for a message of same type
+    if (p_recv->state == RECV_BLOCK && CMP_MASK(recv_mask, request_info->msg_mask)) {
+        // If yes, change state of waiting process to ready and sender to reply block
+        p_recv->state = READY;
+        Cp->state = REPLY_BLOCK;
+
+        // Add the message data and pid of sender to the receiving processes request info
+        p_recv->req_params->msg_ptr_data = request_info->msg_ptr_data;
+        p_recv->req_params->out_pid = Cp->process_id;
+    } else {
+        // If not, sender process goes to send block state
+        MSG *msg = &Messages[Cp->process_id];
+
+        // Save message
+        msg->data = request_info->msg_ptr_data;
+        msg->mask = request_info->msg_mask;
+        msg->receiver = request_info->msg_to;
+
+        Cp->state = SEND_BLOCK;
+    }
+}
+
+void Kernel_Msg_Recv() {
+    // Periodic tasks cannot receive
+    if (Cp->priority == PERIODIC) {
+        OS_Abort(PERIODIC_MSG);
+        return;
+    }
+
+    // Check if there is a message waiting for us
+    PID sender_pid;
+    MSG *msg = NULL;
+    for (sender_pid = 0; sender_pid < MAXTHREAD; sender_pid += 1) {
+        if (Messages[sender_pid].receiver == Cp->process_id) {
+            msg = &Messages[sender_pid];
+            break;
+        }
+    }
+
+    // Check if there is a message waiting for this process
+    if (msg != NULL && CMP_MASK(request_info->msg_mask, msg->mask)) {
+        // If yes, change state to ready and set msg data on Cp's request info
+        request_info->msg_ptr_data = msg->data;
+        request_info->out_pid = sender_pid;
+
+        Cp->state = READY;
+
+        // Sender process now waiting for reply
+        PD *sender = &Process[sender_pid];
+        sender->state = REPLY_BLOCK;
+
+        // Remove data from Messages
+        msg->data = NULL;
+        msg->receiver = MAXTHREAD;
+    } else {
+        // If not, set process to receive block state
+        Cp->state = RECV_BLOCK;
+    }
+}
+
+void Kernel_Msg_Rply() {
+    // Periodic tasks cannot reply
+    if (Cp->priority == PERIODIC) {
+        OS_Abort(PERIODIC_MSG);
+        return;
+    }
+
+    // Check if process id is valid
+    if (!VALID_ID(request_info->msg_to)) {
+        OS_Abort(INVALID_REQ_INFO);
+        return;
+    }
+
+    PD *p_recv = &Process[request_info->msg_to];
+
+    // Check if process replying to is in reply block state
+    if (p_recv->state == REPLY_BLOCK) {
+        p_recv->state = READY;
+
+        // I don't think this will work
+        p_recv->req_params->msg_data = request_info->msg_data;
```

Comment:
Don't think it will work because of where the data is allocated? (<-- Assuming that) This could be an issue since the request_info might fall out of scope before the receiver gets the CPU back. That could happen if we dispatch back to the caller here, case if caller has higher priority

Reply:
But then again, this might actually be okay. I thought `msg_data` was a pointer, but this is a value copy into the blocked receiver's msg info, so when they get CPU back - whenever that is - the value will still be stored on their stack

Reply:
Yeah, I think it should be okay but I'm still not super convinced. I think I'll leave it in for now and just test it a bunch.